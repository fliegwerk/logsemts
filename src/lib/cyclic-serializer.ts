/**
 * @author Pablo Klaschka
 *
 * @packageDocumentation
 *
 * @example ```ts
 * let a = {};
 * let b = { a:a }
 * a['b'] = a;
 *
 * JSON.stringify(a); // throws Error because of cyclic dependency
 *
 * // serialize a, which returns "instructions" for re-building its structure
 * const serialized = serialize(a);
 *
 * // rebuild a's structure from "instructions"
 * const deserialized = deserialize(serialized);
 *
 * console.log(deserialized); // logs the "correct" structure
 * ```
 */

/**
 * A  JSON-based transport format with support for
 * - cyclic dependencies: `a['b']['a'] === a`
 * - `RegExp`
 * - `Symbol`
 * - `BigInt`
 * - `Date`
 *
 * ### Special cases
 * - functions get converted to strings (using their `.toString()` method)
 * - Regular expressions (`RegExp`) get converted to strings
 * - Symbols get converted to strings
 *
 * @see {@link serialize}
 * @see {@link deserialize}
 *
 * @example ```ts
 * let a = {};
 * let b = { a:a }
 * a['b'] = a;
 *
 * JSON.stringify(a); // throws Error because of cyclic dependency
 *
 * // serialize a, which returns "instructions" for re-building its structure
 * const serialized = serialize(a);
 *
 * // rebuild a's structure from "instructions"
 * const deserialized = deserialize(serialized);
 *
 * console.log(deserialized); // logs the "correct" structure
 * ```
 */
export interface SerializedTransportFormat {
	/**
	 * A memory pool for values the pointers in {@link pointers} can point to
	 */
	pool: any[];
	/**
	 * Pointers (indexes) of values stored in the {@link pool} in the order they can get built in.
	 *
	 * Separations are achieved using the `->` string.
	 *
	 * The pointer to the serialized value is `''`, meaning `pointers['']` is the serialized element's index in the
	 * {@link pool}.
	 *
	 * @example ```js
	 * // the serialized value points to an object (`s`) that is at index 0 in the pool
	 * // `s['a']` has the value of an array stored at index 1 in the pool.
	 * // this array has `s` as its only member, i.e., `s['a'][0] === s`:
	 * { '': 0, '->a': 1, '->a->0': 1 }
	 * ```
	 */
	pointers: { [key: string]: number };
	/**
	 * The tye of the value in the {@link pool} with the same index.
	 *
	 * Can be either a value "returned" by the `typeof` operator or a special value for supported types:
	 * - `regexp` - a `RegExp` object
	 */
	types: (
		| 'undefined'
		| 'object'
		| 'boolean'
		| 'number'
		| 'string'
		| 'symbol'
		| 'function'
		| 'bigint'
		| 'regexp'
		| 'Date'
	)[];
}

/**
 * Serialize a value to the {@link SerializedTransportFormat} specifications
 *
 * @param value the value that gets serialized
 * @return the serialized JSON string
 *
 * @see {@link deserialize}
 *
 * @example ```ts
 * let a = {};
 * let b = { a:a }
 * a['b'] = a;
 *
 * JSON.stringify(a); // throws Error because of cyclic dependency
 *
 * // serialize a, which returns "instructions" for re-building its structure
 * const serialized = serialize(a);
 *
 * // rebuild a's structure from "instructions"
 * const deserialized = deserialize(serialized);
 *
 * console.log(deserialized); // logs the "correct" structure
 * ```
 */
export function serialize(value: any) {
	// Create an empty instructions set that'll get modified
	let instructions: SerializedTransportFormat = {
		pool: [],
		pointers: {},
		types: []
	};

	serializeItem('', value, instructions); // serialize the value into ''

	/**
	 * replace values containing other values (objects and arrays) with their empty counterparts as they'll get rebuilt from the instructions
	 * @param value the value
	 */
	const replacer = (value: any) => {
		if (Array.isArray(value)) return [];
		else if (typeof value === 'object' && value !== null) return {};
		else return value;
	};

	instructions.pool = instructions.pool.map(replacer);

	// Finally, the instructions can be JSON-serialized and returned:
	return JSON.stringify(instructions);
}

/**
 * Deserialize a JSON string representing a {@link SerializedTransportFormat}.
 *
 * @param serialized JSON string with a {@link SerializedTransportFormat} format
 *
 * @see {@link serialize}.
 */
export function deserialize(serialized: string) {
	const instructions: SerializedTransportFormat = JSON.parse(serialized);

	instructions.pool = instructions.pool.map((value: any, index: number) => {
		if (instructions.types[index] === 'regexp') {
			const match = value.match(/^\/(.*)\/([gimy]*)/);
			if (match) {
				const [, pattern, flags] = match;
				return new RegExp(pattern, flags);
			} else {
				throw new Error('Invalid regular expression string: ' + value);
			}
		} else if (instructions.types[index] === 'Date') {
			return new Date(value)
		} else if (instructions.types[index] === 'bigint') {
			return BigInt(value);
		} else {
			return value;
		}
	});

	let builder: any = {}; // create an empty object in which the value can get built

	for (let pointer of Object.keys(instructions.pointers)) {
		const steps = pointer.split('->');

		const toSet = steps.pop();

		let current: any = builder;
		steps.forEach(step => {
			current = current[step];
		});

		current[toSet!] = instructions.pool[instructions.pointers[pointer]];
	}

	return builder[''];
}

/**
 * Serialize a value (some sub-item of the main serialized value) and write it into the `instructions`.
 *
 * @param path the path of the property that gets serialized, separated with `->`
 * @param value the value that should get serialized
 * @param instructions the instructions object
 *
 * @internal
 */
function serializeItem(
	path: string,
	value: any,
	instructions: SerializedTransportFormat
) {
	/**
	 * The index of the current value in the pool
	 *
	 * `-1` if it hasn't been added to the pool, yet
	 */
	let poolIndex = instructions.pool.indexOf(value);

	/**
	 * add pointer of path to the value's position in the pool (that gets created if not existent)
	 */
	function addModifierForValue(
		stringify: boolean = false,
		customType?: 'regexp' | 'Date'
	) {
		const valueExistsInPool = poolIndex >= 0;
		let index = valueExistsInPool
			? poolIndex
			: instructions.pool.push(stringify ? value.toString() : value) - 1;
		instructions.pointers[path] = index;
		if (!valueExistsInPool) {
			instructions.types[index] = customType || typeof value;
		}
	}

	if (
		typeof value === 'symbol' ||
		typeof value === 'bigint' ||
		typeof value === 'function'
	) {
		addModifierForValue(true);
	} else if (value instanceof RegExp) {
		addModifierForValue(true, 'regexp');
	} else if (value instanceof Date) {
		addModifierForValue(true, 'Date')
	} else if (Array.isArray(value)) {
		addModifierForValue();

		// add all array items as `path->${indexInArray}` to the serialization
		if (poolIndex < 0)
			value.forEach((item, index) => {
				serializeItem(path + `->${index}`, item, instructions);
			});
	} else if (typeof value === 'object') {
		addModifierForValue();

		// add all object properties as `path->${propertyName}` to the serialization
		if (poolIndex < 0 && value !== null)
			Object.keys(value).forEach(propertyName => {
				serializeItem(
					path + `->${propertyName}`,
					value[propertyName],
					instructions
				);
			});
	} else if (
		typeof value === 'string' ||
		typeof value === 'number' ||
		typeof value === 'boolean'
	) {
		addModifierForValue();
	}
}
