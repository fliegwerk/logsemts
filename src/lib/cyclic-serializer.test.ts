import {deserialize, serialize} from "./cyclic-serializer";

describe('serialize()', () => {
    it('should serialize cyclic properties', () => {
        let a: any = {};
        a['b'] = {a};

        expect(a).toMatchSnapshot('a');
        const serialized = JSON.parse(serialize(a));

        expect(serialized).toMatchSnapshot('serialized a');

        expect(serialized.pool).toHaveLength(2);
        expect(serialized.pool).toContainEqual({});

        expect(serialized.modifiers).toHaveProperty('', 0)
        expect(serialized.modifiers).toHaveProperty('->b', 1)
        expect(serialized.modifiers).toHaveProperty('->b->a', 0)
    })

    it('should serialize strings', () => {
        expect(serialize('abc')).toMatchSnapshot()
    })

    it('should serialize numbers', () => {
        expect(serialize(3)).toMatchSnapshot()
    })

    it('should serialize booleans', () => {
        expect(serialize(true)).toMatchSnapshot()
        expect(serialize(false)).toMatchSnapshot()
    })

    it('should serialize arrays', () => {
        expect(serialize(['abc', 'cde'])).toMatchSnapshot()
    })

    it('should serialize two references to the same array', () => {
        let arr = [1,2];
        const serialized = serialize([arr, arr]);
        expect(serialized).toMatchSnapshot()
        expect(typeof serialized).toBe('string');
    })

    it('should serialize regular expressions to strings', () => {
        const serialized = serialize(/.*/g);
        expect(serialized).toMatchSnapshot()
        expect(typeof serialized).toBe('string');
    })

    it('should serialize functions to strings', () => {
        const serialized = serialize(() => console.log('Hello World!'));
        expect(serialized).toMatchSnapshot()
        expect(typeof serialized).toBe('string');
    })

    it('should serialize symbols to strings', () => {
        const serialized = serialize(Symbol('abc'));
        expect(serialized).toMatchSnapshot()
        expect(typeof serialized).toBe('string');
    })

    it('should serialize null', () => {
        const serialized = serialize(null);
        expect(serialized).toMatchSnapshot()
        expect(typeof serialized).toBe('string');
    })

    it('should serialize undefined', () => {
        const serialized = serialize(undefined);
        expect(serialized).toMatchSnapshot()
        expect(typeof serialized).toBe('string');
    })
})

describe('deserialize()', () => {
    it('should deserialize cyclic properties', () => {
        const serialized = {
            "modifiers": {
                "": 0,
                "->b": 1,
                "->b->a": 0,
            },
            "pool": [
                {},
                {},
            ],
        }

        const deserialized = deserialize(JSON.stringify(serialized));
        expect(deserialized).toMatchSnapshot('deserialized');
        expect(deserialized['b']['a']).toBe(deserialized);
    })

    it('should deserialize undefined', () => {
        const serialized = {pool:[], modifiers: {}};
        expect(deserialize(JSON.stringify(serialized))).toBeUndefined();
    })
})
