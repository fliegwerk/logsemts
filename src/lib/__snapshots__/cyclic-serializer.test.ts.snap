// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serialize() should serialize a Date 1`] = `"{\\"pool\\":[\\"2000-01-31T23:00:00.000Z\\"],\\"pointers\\":{\\"\\":0},\\"types\\":[\\"Date\\"]}"`;

exports[`serialize() should serialize a regular expression 1`] = `"{\\"pool\\":[\\"/fwjeofj/g\\"],\\"pointers\\":{\\"\\":0},\\"types\\":[\\"regexp\\"]}"`;

exports[`serialize() should serialize arrays 1`] = `"{\\"pool\\":[[],\\"abc\\",\\"cde\\"],\\"pointers\\":{\\"\\":0,\\"->0\\":1,\\"->1\\":2},\\"types\\":[\\"object\\",\\"string\\",\\"string\\"]}"`;

exports[`serialize() should serialize booleans 1`] = `"{\\"pool\\":[true],\\"pointers\\":{\\"\\":0},\\"types\\":[\\"boolean\\"]}"`;

exports[`serialize() should serialize booleans 2`] = `"{\\"pool\\":[false],\\"pointers\\":{\\"\\":0},\\"types\\":[\\"boolean\\"]}"`;

exports[`serialize() should serialize functions to strings 1`] = `"{\\"pool\\":[\\"function () { return console.log('Hello World!'); }\\"],\\"pointers\\":{\\"\\":0},\\"types\\":[\\"function\\"]}"`;

exports[`serialize() should serialize null 1`] = `"{\\"pool\\":[null],\\"pointers\\":{\\"\\":0},\\"types\\":[\\"object\\"]}"`;

exports[`serialize() should serialize numbers 1`] = `"{\\"pool\\":[3],\\"pointers\\":{\\"\\":0},\\"types\\":[\\"number\\"]}"`;

exports[`serialize() should serialize strings 1`] = `"{\\"pool\\":[\\"abc\\"],\\"pointers\\":{\\"\\":0},\\"types\\":[\\"string\\"]}"`;

exports[`serialize() should serialize symbols to strings 1`] = `"{\\"pool\\":[\\"Symbol(abc)\\"],\\"pointers\\":{\\"\\":0},\\"types\\":[\\"symbol\\"]}"`;

exports[`serialize() should serialize two references to the same array 1`] = `"{\\"pool\\":[[],[],1,2],\\"pointers\\":{\\"\\":0,\\"->0\\":1,\\"->0->0\\":2,\\"->0->1\\":3,\\"->1\\":1},\\"types\\":[\\"object\\",\\"object\\",\\"number\\",\\"number\\"]}"`;

exports[`serialize() should serialize undefined 1`] = `"{\\"pool\\":[],\\"pointers\\":{},\\"types\\":[]}"`;
